server.go

```diff
@@ -243,6 +245,7 @@ func NewFromConfig(logger *logrus.Logger, conf Config) (*Server, error) {
 	if err != nil {
 		return ret, err
 	}
+	ret.derivedMetrics = metricSink
```

Comment:
I'm having a hard time following the naming of these now.

sssfmetrics.NewMetricExtractionSink creates a span sink that takes in `ssf.SSFSpan` values, transforms thenm to `ssf.SSFSample`, and then reports those to the _metric_workers in Veneur.

Now, we're using that same metrics extraction sink to initialize the SignalFX sink (`NewSignalFxSink`). When the SignalFX sink encounters a service check, it will generate an SSF metric, and send it to the metrics extraction sink we're providing here.

However, we *also* pass that metricSink value to the _spanSinks_, which puts it on the same level as Lightstep, Kafka, etc. - one of the end destinations that Veneur ultimately passes spans off to. It does that so that we can extract metrics (`samplers.UDPMetric`) and send those to the regular metrics workers in Veneur, which will process them and send them to all metrics sinks (including SignalFx). 

This means our dataflow chart technically has a loop in it. I don't think it's a problem in practice, because SignalFx only uses `derivedMetrics` when it ingests a service check, and it never *outputs* a service check (because the `ssf.SSFSample` doesn't even support that, though it *does* support a Status type).

In other words, `DerivedMetricsProcessor` should *only* be used to transform types which SSF does not support natively into SSF-native values. If anyone provides an implementation that transforms types which SSF *does* support natively into SSF-native values, then we'll open up the possibility of write-amplification. In fact, we have a slightly stricter requirement than that: the values passed through a `DerivedMetricsProcessor` *must* originate from an external, non-Veneur source; we can't have them create SSF-native values even indirectly. (For example, consider the situation: Veneur emits a metric `foo.bar` to itself. Somewhere else internally, we happen to emit a service check every time we see `foo.bar` as a metric. This then results in us invoking a DerivedMetricsProcessor on the service check which was derived from `foo.bar`, which creates another internal metric).

That's a little contrived as an example because Veneur has no reason to emit service checks itself, ever, but the DerivedMetricsProcessor is written to be general-purpose, not limited to service checks.

So, a few questions:


1. Is the above all correct?

2. Do we need to store `derivedMetrics` on the server at all, or can we just pass `metricSink` directly to `signalfx.NewSignalFxSink`?

3. Can we rename the local value `metricSink` to something a bit more descriptive, given that it has multiple roles internally? I know you didn't name metricSink as part of this PR, but it's only now that we're using it in two different ways that I'm encountering this confusion

4. Can we update the docs around `DerivedMetricsProcessor` to clarify that (a) it should only ever be used for converting types that are not supported natively by SSF, and (b) It should only ever be called on values that originate from "beyond" Veneur. (Thankfully, Veneur uses SSF for all of its own internal metrics reporting, so there's no possibility of us accidentally emitting a type that SSF doesn't support from within Veneur).

Reply:
I'm going to sync up with you directly to discuss this in higher bandwidth communication.